### PR TITLE
Issue #10309 - Duplicate `X-Powered-By` headers

### LINKED ISSUE #10309: Jetty 12: X-Powered-By header is added 2 times (if enabled)
**Jetty version(s)**
Jetty 12

**Jetty Environment**
core

**Description**
If you enable `HttpConfiguration#setSendXPoweredBy(true)`, then Jetty returns the header 2 times (at least with HTTP/1.1):

```
> curl -I http://localhost:8081/
HTTP/1.1 200 OK
X-Powered-By: Jetty(12.0.0)
Date: Mon, 14 Aug 2023 18:38:26 GMT
Content-Type: text/plain;charset=utf-8
Content-Length: 46
X-Powered-By: Jetty(12.0.0)
```

### PR DESCRIPTION
**Pull request by joakime**
Turned out that both `HttpGenerator` and `HttpChannelState` were adding the `X-Powered-By` and `Server` headers.
This removes the addition of the `X-Powered-By` header from `HttpChannelState`.

### CODE REVIEW

**Review by gregw — CHANGES_REQUESTED** [review]
I think we need to do this the other way around.  `HttpChannelState` will add them for all protocols, whilst HttpGenerator only does them for HTTP/1

**Comment by joakime** [comment]
> I think we need to do this the other way around. `HttpChannelState` will add them for all protocols, whilst HttpGenerator only does them for HTTP/1

This will be a bigger change, but sure, I'll do it that way.

**Comment by uschindler** [comment]
Hi,
I like sending the headers using the normal `HttpFields` mechanism instead of printed directly to the wire protocol. This allows to modify or overwrite them in the handler code. I was just curious if those fields are restored if you do `Response#reset()`.

**Comment by sbordet** [comment]
@joakime can you write 1 more test to verify what @uschindler says above?

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java [hunk 1]
```diff
@@ -194,6 +196,80 @@ public boolean handle(Request request, Response response, Callback callback)
         assertThat(response.get(HttpHeader.LOCATION), is("/somewhere/else"));
     }
 
+    @Test
+    public void testXPoweredByDefault() throws Exception
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:200]: I'd like to see a test that calls `Response.reset()` and verifies those default headers are set again.

**Review by sbordet — DISMISSED** [review]

**Comment by gregw** [comment]
> I was just curious if those fields are restored if you do `Response#reset()`

@uschindler very good question.  This is an issue that I've been pondering for some time and have not come up with a good answer for.... so was hoping nobody noticed :)

However, I have a few ideas how it can be done.  @joakime I'll make a branch with the rough idea I'd thought about but not implemented.... it probably will need some work/testing/throwing-out-and-starting-again

**Comment by gregw** [comment]
@joakime this is my thought bubble for how to freeze headers: https://github.com/eclipse/jetty.project/compare/jetty-12.0.x...fix/10310/freezeHttpFields?expand=1

**Comment by gregw** [comment]
@joakime do you want me to merge my thought bubble here or to make it a separate PR

**Comment by joakime** [comment]
> @joakime do you want me to merge my thought bubble here or to make it a separate PR

Separate PR please

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java (hunk 1 above)
Comment (joakime) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:200]: The place where those headers are added is now in `HttpChannelState.onRequest`.
A call to `Response.reset()` erases those headers, and another `HttpChannelState.onRequest()` needs to occur.

This is true for `Server`, `X-Powered-By`, and `Date` headers.

**Comment by joakime** [comment]
Lets get this PR as-is merged, then we can focus on the improvements that #10339 bring to the table (answering reset and persistent headers) as a separate PR.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:200]: I've got the reset test in PR #10339

**Review by gregw — APPROVED** [review]

**Pull request merged**